lib/customisation.nix

```diff
@@ -296,4 +296,32 @@ rec {
       };
     in self;
 
+
+
+  /* Apply a function to an attrset or potential args, automatically filtering
+     the set to those the function can accept.
+     NOTE: This is strictly meant to be used with functions that accept attrsets
+     as their argument, and is best suited for use with `setFunctionArgs'.
+
+     nix-repl> lib.apply ( { x, y }: x + y ) { x = 1; y = 2; z = -1; }
+     3
```

Comment:
This has a very similar effect to

 - https://github.com/NixOS/nixpkgs/pull/194514

Do we need the body to be strict in the attribute names? If not, `lazyFunction` is strictly more useful, pardon the pun.

Reply:
No the arguments don't need to be split, not sure why I wrote it this way

Reply:
Agreed that this has some overlap with `lazyFunction` ( I like your style, https://github.com/aakropotkin/ak-nix/blob/7242b7d7c49037700def339ea232d594b67dcd7c/lib/funk.nix#L346 there's no denying that we've got similar taste ); but my goal with this routine was intentionally limited to filtering a set of `auto` args rather than making them lazy ( which is just a side effect ).

If you've got any recommendations to avoid stepping on each others toes let me know. We wouldn't want to add redundant routines and you definitely beat me to the punch with this PR in October.

Reply:
I don't think yours actually makes it lazy though?

```
nix-repl> lib.fix (lib.applyArgs ({ a, b }: { a = b; b = 1; }) )
error: infinite recursion encountered
```

Maybe this looks like making up requirements, which in a way it is, but I'm concerned about function sprawl. As in [the Fairbairn threshold](https://wiki.haskell.org/Fairbairn_threshold).

Reply:
It's a real tragedy that we have an incentive to inline frequently called functions like this to gain better performance.

Reply:
I mean, I wouldn't expect that `lib.applyArgs` snippet to work because you've just got a lambda there. Lets review the definition:
```
  applyArgs = f: args: let
    fn = if lib.isFunction f then f else import f;
  in fn ( builtins.intersectAttrs ( lib.functionArgs fn ) args );
```

The note I have about "this is only meant for functions that accept attrsets" was alluding to this a bit.

I'd describe this routine as more of "useful idiom"/short-hand rather than a generalized "works everywhere, do what I mean" utility ( we've got more robust routines for those situations anyways ).

Reply:
It's not just a lambda, but equivalent to

```
nix-repl> let r = lib.applyArgs ({ a, b }: { a = b; b = 1; }) r; in r
error: infinite recursion encountered
```

Reply:
This is a good observation, because it's strict we avoid a lot of overhead when a function is used frequently, but you get the usual constraints for strict evaluations.

`applyArgsStrict` and `applyArgsLazy` could make this more obvious to users, but it doesn't quite roll off the tongue the same way.